### PR TITLE
feature: extend cri apis for special needs

### PR DESCRIPTION
**Pull request by Starnop**
Signed-off-by: Starnop <starnop@163.com>

### Ⅰ. Describe what this PR did
At present, the CRI interface of Kubernetes cannot meet the customized development of Kubelet. Therefore, it is necessary to provide the required functions of Kubelet by extending CRI.

### CODE REVIEW

File: cri/apis/cri/runtime/v1alpha2/api.proto [hunk 1]
```diff
@@ -0,0 +1,1281 @@
+// To regenerate api.pb.go run hack/update-generated-runtime.sh
```
Comment (YaoZengzeng) [inline cri/apis/cri/runtime/v1alpha2/api.proto]: `cri/apis/cri/v1alpha1` ---> `cri/apis/v1alpha1`
`cri/apis/cri/runtime/v1alpha2` ---> `cri/apis/v1alpha2`

make the directory structure more clear

File: cri/v1alpha2/cri.go [hunk 2]
```diff
@@ -482,9 +483,11 @@ func (c *CriManager) CreateContainer(ctx context.Context, r *runtime.CreateConta
 			OpenStdin: config.Stdin,
 			StdinOnce: config.StdinOnce,
 			Tty:       config.Tty,
+			DiskQuota: resources.GetDiskQuota(),
 		},
 		HostConfig: &apitypes.HostConfig{
-			Binds: generateMountBindings(config.GetMounts()),
+			Binds:     generateMountBindings(config.GetMounts()),
+			Resources: parseResourceFromAPI(resources),
```
Comment (YaoZengzeng) [inline cri/v1alpha2/cri.go]: rename `parseResourceFromAPI` to `parseResourcesFromAPI` may more consistent :)

File: cri/v1alpha2/cri_utils.go [hunk 3]
```diff
@@ -791,3 +792,152 @@ func (c *CriManager) attachLog(logPath string, containerID string) error {
...
+	}
+	return throttleDevice
+}
+
+// parseUlimitFromAPI parse Ulimit from runtime.Ulimit to apitypes.Ulimit
+func parseUlimitFromAPI(runtimeUlimit []*runtime.Ulimit) []*apitypes.Ulimit {
+	if runtimeUlimit == nil {
+		return nil
+	}
+	var ulimit []*apitypes.Ulimit
```
Comment (YaoZengzeng) [inline cri/v1alpha2/cri_utils.go]: Please make sure how to use slice in golang.

File: cri/v1alpha1/cri_utils.go [hunk 4]
```diff
@@ -940,3 +941,158 @@ func blkrdev(device string) (uint64, error) {
...
+	for _, v := range apitypesWeightDevices {
+		weightDevices = append(weightDevices, &runtime.WeightDevice{
+			Path:   v.Path,
+			Weight: uint32(v.Weight),
+		})
+	}
+	return weightDevices
+}
+
+// parseThrottleDeviceFromAPI parse ThrottleDevice from runtime.WeightDevice to apitypes.WeightDevice
```
Comment (YaoZengzeng) [inline cri/v1alpha1/cri_utils.go]: The `WeightDevice` in comment should be `ThrottleDevice`?

File: cri/v1alpha1/cri_utils.go [hunk 5]
```diff
@@ -940,3 +941,158 @@ func blkrdev(device string) (uint64, error) {
...
+	for _, v := range runtimeThrottleDevices {
+		throttleDevices = append(throttleDevices, &apitypes.ThrottleDevice{
+			Path: v.GetPath(),
+			Rate: v.GetRate(),
+		})
+	}
+	return throttleDevices
+}
+
+// parseThrottleDeviceFromPouch parse ThrottleDevice from apitypes.WeightDevice to runtime.WeightDevice
```
Comment (YaoZengzeng) [inline cri/v1alpha1/cri_utils.go]: The `WeightDevice` in comment should be `ThrottleDevice`?

Comment (Starnop) [reply cri/v1alpha1/cri_utils.go]: I'm sorry for my carelessness. it has been corrected.

File: cri/apis/v1alpha1/api.pb.go [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2017 The Kubernetes Authors.
+Copyright YEAR AUTHORS.
```
Comment (fuweid) [inline cri/apis/v1alpha1/api.pb.go]: YEAR AUTHORS?

File: cri/v1alpha1/cri_utils.go [hunk 7]
```diff
@@ -940,3 +941,137 @@ func blkrdev(device string) (uint64, error) {
 	}
 	return 0, fmt.Errorf("cannot get stat of device %s", device)
 }
+
+// CRI extension related tool functions.
+
+// parseResourceFromAPI parse Resources from runtime.LinuxContainerResources to apitypes.Resources
+func parseResourcesFromAPI(runtimeResources *runtime.LinuxContainerResources) apitypes.Resources {
```
Comment (fuweid) [inline cri/v1alpha1/cri_utils.go]: s/parseResoucesFromAPI/pbLinuxContainerResourcesToAPIResouces/g WDYT?

File: cri/v1alpha1/cri_utils.go [hunk 8]
```diff
@@ -940,3 +941,137 @@ func blkrdev(device string) (uint64, error) {
...
+		})
+	}
+	return
+}
+
+// parseVolumesFromPouch parse Volumes from map[string]interface{} to map[string]*runtime.Volume
+func parseVolumesFromPouch(containerVolumes map[string]interface{}) map[string]*runtime.Volume {
+	volumes := make(map[string]*runtime.Volume)
+	for k, v := range containerVolumes {
+		volumes[k] = v.(*runtime.Volume)
```
Comment (fuweid) [inline cri/v1alpha1/cri_utils.go]: always `(*runtime.Volume)` ? nil or other type?

Comment (Starnop) [reply cri/v1alpha1/cri_utils.go]: Done. thanks for your comment.

File: cri/v1alpha1/cri.go [hunk 9]
```diff
@@ -728,7 +734,9 @@ func (c *CriManager) ContainerStatus(ctx context.Context, r *runtime.ContainerSt
 		Message:     message,
 		Labels:      labels,
 		Annotations: annotations,
-		// TODO: LogPath.
+		LogPath:     container.LogPath,
```
Comment (YaoZengzeng) [inline cri/v1alpha1/cri.go]: The `container.LogPath` is not the log path of cri container

I'll fix it in my next PR, please leave it empty now :)

File: hack/protoc.sh [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        --proto_path="${DIR}/vendor" \
+        --gogo_out=plugins=grpc:"${API_ROOT}" "${API_ROOT}/api.proto"
+    # Update boilerplate for the generated file.
+    cat "${DIR}/hack/boilerplate/boilerplate.go.txt" "${API_ROOT}/api.pb.go" > "${API_ROOT}/tmpfile" \
+        && mv "${API_ROOT}/tmpfile" "${API_ROOT}/api.pb.go"
+    gofmt -l -s -w "${API_ROOT}/api.pb.go"
+}
+
+main(){
+    generateproto
```
Comment (fuweid) [inline hack/protoc.sh]: API_ROOT="${DIR}/cri/apis/v1alpha1" generateproto
API_ROOT="${DIR}/cri/apis/v1alpha2" generateproto

**Comment by YaoZengzeng** [comment]
I've used this PR to do node-e2e test, it works fine.

LGTM

**Pull request merged**